packages/desktop-client/src/components/reports/spreadsheets/master-spreadsheet.js

```diff
@@ -0,0 +1,432 @@
+          premadeLabel: label,
+        });
+      }
+      if (expense !== 0) {
+        res.expenses.push({
+          x,
+          y: integerToAmount(expense),
+          q: integerToCurrency(expense),
+          premadeLabel: label,
+        });
+      }
+
+      return res;
+    },
+    { expenses: [], income: [] },
+  );
+
+  console.log(totalExpensesCat);
+  console.log(totalIncomeCat);
+
+  const dates = isConcise
+    ? monthUtils.rangeInclusive(
+        monthUtils.getMonth(start),
+        monthUtils.getMonth(end),
+      )
+    : monthUtils.dayRangeInclusive(start, end);
+
+  const incomes = index(
+    incomeDate,
+    'date',
+    isConcise ? fromDateRepr : fromDateReprToDay,
+  );
+
+  const expenses = index(
+    expenseDate,
+    'date',
+    isConcise ? fromDateRepr : fromDateReprToDay,
+  );
+
+  let balance = !(isCashFlow || isNetWorth) ? 0 : startingBalance;
+  let totalExpenses = 0;
+  let totalIncome = 0;
+  let debt = startingDebts;
+  let assets = startingAssets;
+  let hasNegative = false;
+  let startNetWorth = debt + assets;
+  let endNetWorth = 0;
+
+  const graphData = dates.reduce(
+    (res, date) => {
+      let income = 0;
+      let expense = 0;
+
+      const last =
+        res.balances.length === 0
+          ? null
+          : res.balances[res.balances.length - 1];
+
+      income = incomes[date] ? incomes[date].amount : 0;
+      expense = expenses[date]
+        ? expenses[date].amount * (isIE && selectList === 'Expense' ? -1 : 1)
+        : 0;
+
+      totalExpenses += expense;
+      totalIncome += income;
+      balance += income + expense;
+
+      debt += expense;
+      assets += income;
+
+      if (balance < 0) {
+        hasNegative = true;
+      }
+
+      const x = d.parseISO(date);
+
+      if (res.length === 0) {
+        startNetWorth = balance;
+      }
+      endNetWorth = balance;
+
+      const change = last ? balance - amountToInteger(last.y) : 0;
+
+      const label = ChangeLabel(
+        income,
+        expense,
+        balance,
+        assets,
+        debt,
+        change,
+        totalExpenses,
+        totalIncome,
+        isCashFlow,
+        isNetWorth,
+        isIE,
+        isConcise,
+        x,
+      );
+
+      res.income.push({ x, y: integerToAmount(income), premadeLabel: label });
+      res.expenses.push({
+        x,
+        y: integerToAmount(expense),
+        premadeLabel: label,
+      });
+      res.change.push({
+        x,
+        y: integerToAmount(expense + income),
+        premadeLabel: label,
+      });
+      res.balances.push({
+        x,
+        y: integerToAmount(balance),
+        premadeLabel: label,
+      });
+      return res;
+    },
+    { expenses: [], income: [], change: [], balances: [] },
+  );
+
+  return {
+    graphData,
+    catData,
+    totalExpenses,
+    totalIncome,
+    hasNegative,
+    netWorth: endNetWorth,
+    totalChanges: endNetWorth - startNetWorth,
+    startNetWorth,
+  };
+}
+
+function ChangeLabel(
```

Comment:
I’m not a huge fan of these components having their body be essentially one big switch statement. Do you think it would make sense to break out each screen into its own component that takes care of rendering the appropriate information itself? Or at least split out the different components for each of these cases to a separate file?

Reply:
I'm not too sure what you mean. Could you give an example? Where would the actual switch happen upon onClick of the report headers?

Reply:
Right now there are components `ChooseChartHeader`, `ChooseChart`, and `ChartExtraColumn` that render different components based on the value of `reportPage`. I’m proposing that instead you have a `<CashFlowReport />`, `<NetWorthReport />`, and `<IncomeExpensesReport />` that may all use common scaffolding components for layout but each “own” the entire content area of the report. That way all the code for each report’s UI would be in one place. Then in `AllReports` you would look at the value of `reportPage` and choose which of those 3 components to render.

Ideally this state would be hooked up to the router so that each kind of report gets its own URL:

```
/reports/net-worth
/reports/cash-flow
/reports/income-expense?tab=totals&type=expense
/reports/income-expense?tab=totals&type=income
/reports/income-expense?tab=totals&type=all
/reports/income-expense?tab=trends&type=expense
/reports/income-expense?tab=trends&type=income
```

Reply:
This sounds like a good idea.

The reason I chose this way was to avoid re-renders and to house all the data on one spreadsheet. I thought it might speed things up and help avoid crashes. Maybe it has the opposite effect? I don't know if my implementation is better or worse?

It sounds like a code organization issue more than anything? Which is fine. Maybe there's more to it? Just trying to learn. Cheers!

Reply:
@j-f1, I've reorganized the code so it follows you logic makes much more sense this way. I haven't implemented the router URLs you mentioned yet. Just wanted to start here.